website/blog/2024-07-26-pyfunc-in-practice/index.md

```diff
@@ -0,0 +1,930 @@
+    predictions = np.array(
+        [self.fitted_models[model].predict(X_processed) for model in model_list]
+    )
+
+    # Apply the specified strategy to combine the model predictions
+    if "average" in strategy:
+        # Calculate the average of predictions from all models
+        return np.mean(predictions, axis=0)
+    elif "weighted" in strategy:
+        # Extract weights from the strategy and normalize them
+        weights = [float(w) for w in loaded_strategy["weights"].iloc[0].split(",")]
+        weights = np.array(weights)
+        weights /= np.sum(weights)  # Ensure weights sum to 1
+
+        # Compute the weighted average of predictions
+        return np.average(predictions, axis=0, weights=weights)
+    else:
+        # Raise an error if an unknown strategy is encountered
+        raise ValueError(f"Unknown strategy: {strategy}")
+```
+
+### Defining a `load context` custom method to initialize the Ensemble Model
+When loading the ensemble model using `mlflow.pyfunc.load_model`, the custom `load_context` method is executed to handle the required model initialization steps before inference.
+
+This initialization process includes:
+
+1. Loading model artifacts, including both the pre-trained models and the preprocessor, using the context object that contains the artifacts references.
+2. Fetching strategies definitions from DuckDB Database.
+
+```python
+def load_context(self, context) -> None:
+    """
+    Loads the preprocessor and models from the MLflow context.
+
+    Args:
+        context: MLflow context object which provides access to saved artifacts.
+    """
+    # Load the preprocessor if its path is specified in the context artifacts
+    preprocessor_path = context.artifacts.get("preprocessor", None)
+    if preprocessor_path:
+        self.preprocessor = joblib.load(preprocessor_path)
+
+    # Load each model from the context artifacts and store it in the fitted_models dictionary
+    for model_name in self.models.keys():
+        model_path = context.artifacts.get(model_name, None)
+        if model_path:
+            self.fitted_models[model_name] = joblib.load(model_path)
+        else:
+            # Print a warning if a model is not found in the context artifacts
+            print(
+                f"Warning: {model_name} model not found in artifacts. Initialized but not fitted."
+            )
+
+    # Reconnect to the DuckDB database to load the strategies
+    conn = duckdb.connect(self.db_path)
+    # Fetch strategies from the DuckDB database into the strategies DataFrame
+    self.strategies = conn.execute("SELECT * FROM strategies").fetchdf()
+    # Close the database connection
+    conn.close()
+```
+
+### Bringing It All Together
+
+Having explored each method in detail, the next step is to integrate them to observe the complete implementation in action. This will offer a comprehensive view of how the components interact to achieve the objectives of the project.
+
+```python
```

Comment:
For this part, we could forego the full example script and instead do a "build your own" with placeholders for each of the methods defined above. Users can then assemble the class in their own environment and run it to see the results.

Reply:
That’s a really interesting idea! Since we’ve added the class skeleton, it’s now easier for users to integrate the methods and customize them. I’ve removed the full example script and instead included a note guiding users to assemble the `EnsembleModel` class using the provided skeleton and add their own logic as needed (they can even expand the functionalities).